src/datarobotx/idp/guard_configurations.py

```diff
@@ -0,0 +1,360 @@
+# Copyright 2024 DataRobot, Inc. and its affiliates.
+#
+# All rights reserved.
+#
+# DataRobot, Inc.
+#
+# This is proprietary source code of DataRobot, Inc. and its
+# affiliates.
+#
+# Released under the terms of DataRobot Tool and Utility Agreement.
+# https://www.datarobot.com/wp-content/uploads/2021/07/DataRobot-Tool-and-Utility-Agreement.pdf
+
+from typing import Any, Dict, List, Literal, Optional, TypedDict, Union
+
+import datarobot as dr
+from datarobot.utils import from_api, to_api, underscorize
+
+from datarobotx.idp.common.hashing import get_hash
+from datarobotx.idp.custom_model_versions import (
+    get_or_create_custom_model_version_from_previous,
+)
+
+
+class Condition(TypedDict):
+    """A condition for a guard configuration."""
+
+    comparand: float
+    comparator: Literal[
+        "greaterThan",
+        "lessThan",
+        "equals",
+        "notEquals",
+        "is",
+        "isNot",
+        "matches",
+        "doesNotMatch",
+        "contains",
+        "doesNotContain",
+    ]
+
+
+class Intervention(TypedDict):
+    """An intervention for a guard configuration."""
+
+    action: Literal["report", "block"]
+    conditions: List[Condition]
+    message: str
+    send_notification: bool
+
+
+_keys_to_remove = [
+    "createdAt",
+    "creatorId",
+    "entityId",
+    "entityType",
+    "orgId",
+    "allowedStages",
+    "additionalConfig",
+    "productionOnly",
+    "id",
+    "isDeployed",
+]
+
+
+def _clean_guard_configurations(guard_config: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
+    cleaned_config = []
+    for config in guard_config:
+        new_config = {}
+        for k, v in config.items():
+            if k in [underscorize(key) for key in _keys_to_remove]:
+                continue
+            if v is not None:
+                new_config[k] = v
+        cleaned_config.append(new_config)
+    return cleaned_config
+
+
+def _get_latest_model_version_id(endpoint: str, token: str, custom_model_id: str) -> str:
+    custom_model = dr.CustomInferenceModel.get(custom_model_id)  # type: ignore
+    if not custom_model.latest_version:
+        raise ValueError("Custom model has no versions")
+
+    latest_version = custom_model.latest_version
+    return str(latest_version.id)
+
+
+def _get_unfrozen_model_version_id(
```

Comment:
maybe rename _get_or_create... for clarity, i was initially confused when reviewing the calling function until i read the source for this function